Update the training pre-requisites (#48)
quantumblacklabs/kedro-examples: kedro-training/docs/01_prerequisites.md

# Training pre-requisites
The Kedro training materials assume a pre-requisite level of technical understanding.

To optimise your experience and learn the most you can from the Kedro training, please review the following before your training session. We provide external resource links for each topic.

- Intermediate Python knowledge

>   - [Basic Python](https://docs.python.org/3/tutorial/)

Reviewer: I would group it like this:

 - Introduction to Python
   - Functions, loops, conditional statements, IO operation
   - Common data structures: lists, dictionaries, tuples
 - Intermediate Python
   - Installing Python packages using `pip`
   - Dependency management with `requirements.txt`
   - Python modules (e.g how to use __init__.py and relative and absolute imports)
   - Some familiarity with Python data science libraries, especially `pandas`, `scikit-learn`, Jupyter Notebook, iPython)
   - Using a virtual environment (Anaconda, venv, pipenv)

    - Functions, loops, conditional statements, IO operation
    - Common data structures: lists, dictionaries, tuples
  - [Installing Python packages using `pip`](https://pip.pypa.io/en/stable/quickstart/) 
  - [Dependency management with `requirements.txt`](https://pip.pypa.io/en/latest/user_guide/#requirements-files)
  - [Python modules](https://docs.python.org/3/tutorial/modules.html) (e.g how to use `__init__.py` and relative and absolute imports)
  - [Some familiarity with Python data science libraries](https://towardsdatascience.com/top-10-python-libraries-for-data-science-cd82294ec266), especially `Pandas`, `scikit-learn`, [Jupyter notebook/labs](https://www.dataquest.io/blog/jupyter-notebook-tutorial/), [iPython](https://www.codecademy.com/articles/how-to-use-ipython))

>   - [Using a virtual environment](https://docs.python.org/3/tutorial/venv.html) (Anaconda, `venv`, `pipenv`)

Reviewer: Why don't you mention `conda` here?

Author: Changed from Anaconda to `conda` :)

# Training pre-requisites
The Kedro training materials assume a pre-requisite level of technical understanding.

To optimise your experience and learn the most you can from the Kedro training, please review the following before your training session. We provide external resource links for each topic.

- [Introduction to Python](https://docs.python.org/3/tutorial/)

>   - Functions, loops, conditional statements, IO operation

Reviewer: ```suggestion
  - Functions, loops, conditional statements and IO operation
```

>   - Common data structures: lists, dictionaries, tuples

Reviewer: ```suggestion
  - Common data structures including lists, dictionaries and tuples
```

- Intermediate Python
  - [Installing Python packages using `pip`](https://pip.pypa.io/en/stable/quickstart/)
  - [Dependency management with `requirements.txt`](https://pip.pypa.io/en/latest/user_guide/#requirements-files)
  - [Python modules](https://docs.python.org/3/tutorial/modules.html) (e.g how to use `__init__.py` and relative and absolute imports)

>   - [Some familiarity with Python data science libraries](https://towardsdatascience.com/top-10-python-libraries-for-data-science-cd82294ec266), especially `Pandas`, `scikit-learn`, [Jupyter notebook/labs](https://www.dataquest.io/blog/jupyter-notebook-tutorial/), [iPython](https://www.codecademy.com/articles/how-to-use-ipython))

Reviewer: ```suggestion
  - [Familiarity with Python data science libraries](https://towardsdatascience.com/top-10-python-libraries-for-data-science-cd82294ec266), especially `Pandas` and `scikit-learn`
  - An understanding of how to use [Jupyter Notebook/Lab](https://www.dataquest.io/blog/jupyter-notebook-tutorial/) and [iPython](https://www.codecademy.com/articles/how-to-use-ipython))
```

  - [Using a virtual environment](https://docs.python.org/3/tutorial/venv.html) (we recommend using `conda`, but you can also use `venv` or `pipenv`)
- [Basic YAML syntax](https://yaml.org/)

> - [Working with the command line](https://tutorial.djangogirls.org/en/intro_to_command_line/) (aka. cmd, CLI, prompt, console or terminal)

Reviewer: ```suggestion
- [Working with the command line](https://tutorial.djangogirls.org/en/intro_to_command_line/) (also known as cmd, CLI, prompt, console or terminal)
```

  -  `cd` to navigate directories
  -  `ls` to list files and directories
  -  [Executing a command and Python program from a command line](https://realpython.com/run-python-scripts/#how-to-run-python-scripts-using-the-command-line)

> The following lists technologies that Kedro integrates with, but it won't cause any issues if you are not yet familiar with them.

Reviewer: ```suggestion
The following lists software that Kedro integrates with; however, during training these are optional requirements.
```

- [Version Control with Git](https://git-scm.com/doc)

> - Cloud storage ([S3](https://aws.amazon.com/s3/), [Azure Blob](https://azure.microsoft.com/en-gb/services/storage/blobs/), [GCS](https://cloud.google.com/storage))

Reviewer: ```suggestion
- Cloud storage ([S3](https://aws.amazon.com/s3/), [Azure Blob](https://azure.microsoft.com/en-gb/services/storage/blobs/) and [GCS](https://cloud.google.com/storage))
```

- [PySpark](https://spark.apache.org/docs/latest/api/python/index.html)

> - [Docker](https://docs.docker.com/) for managing deployment

Reviewer: ```suggestion
- [Docker](https://docs.docker.com/) as a deployment option
```

- [Airflow](https://airflow.apache.org/docs/stable/tutorial.html) for scheduling pipeline execution

# Checklist
Please use this checklist to make sure you have everything necessary to participate in the Kedro training.

> - [ ] You have [Python 3 (either 3.6, 3.7 or 3.8)](https://www.python.org/downloads/) installed in your laptop.

Reviewer: ```suggestion
- [ ] You have [Python 3 (either 3.6, 3.7 or 3.8)](https://www.python.org/downloads/) installed on your laptop
```

- [ ] You have Anaconda or an [alternative](https://github.com/quantumblacklabs/kedro-examples/blob/master/kedro-training/docs/02_virtual-environment.md) virtual environment manager

> - [ ] You have [an editor](#text-editors-and-ides) installed for writing Python code

Reviewer: ```suggestion
- [ ] You have [a code editor](#text-editors-and-ides) installed for writing Python code
```

> - [ ] You have a Terminal installed

Reviewer: How do they check that a Terminal is installed?

Author: I've added a new subsection for command line and added the anchor link in here :)

> Having completed the above checklist, make sure that you are able to execute the following commands in your Terminal

Reviewer: ```suggestion
Having completed the above checklist, make sure that you are able to execute the following commands in your Terminal:
```

- [ ]  `python --version` or `python3 --version` returns a correct Python version (either 3.6, 3.7 or 3.8).

> - [ ] `git clone https://github.com/quantumblacklabs/kedro-examples` downloads this training Github repository (or [manually download a zip file from the link](https://stackoverflow.com/questions/2751227/how-to-download-source-in-zip-format-from-github)).

Reviewer: This makes an assumption that they've done the `git` work. Rather just tell tell them to download the repo, so they all have one way to do it.

```suggestion
- [ ] Download the [Kedro training repository](https://github.com/quantumblacklabs/kedro-examples) by following [these instructions](https://stackoverflow.com/questions/2751227/how-to-download-source-in-zip-format-from-github).
```

> - [ ] `conda create --name=training python=3.6 -y && conda activate training` creates an virtual environment called `training` and activates the environment.

Reviewer: This assumes they have Anaconda installed and are not using `venv`... How could this be better written? Is it possible to just link them to setup instructions?

Author: I would keep `conda create --name=training python=3.6 -y && conda activate training` , but also liked the setup instruction for venv and pipenv :)

- [ ] `pip install kedro` installs Kedro in your conda environment.

- [ ]  `kedro --version` shows [the latest Kedro version](https://pypi.org/project/kedro/).

If you are able to complete all of the above, you are ready for the training! In case you have any problems or questions in any of the above checklist, please contact an instructor and resolve the issues before the training.

# Installation prerequisites

Kedro supports macOS, Linux and Windows (7 / 8 / 10 and Windows Server 2016+). If you encounter any problems on these platforms please engage Kedro community support on [Stack Overflow](https://stackoverflow.com/questions/tagged/kedro).

## Python

> Kedro supports Python 3.6, 3.7 and 3.. We recommend using [Anaconda](https://www.anaconda.com/download) (Python 3.7 version) to install Python packages. However, if Anaconda is not preferred then this [tutorial](https://realpython.com/installing-python/) can help you with installation and setup.

Reviewer: ```suggestion
Kedro supports Python 3.6, 3.7 and 3.8. We recommend using [Anaconda](https://www.anaconda.com/download) (Python 3.7 version) to install Python packages. However, if Anaconda is not preferred then this [tutorial](https://realpython.com/installing-python/) can help you with installation and setup.
```

# Training pre-requisites

> This Kedro training materials assume the following pre-requisites for your technical understanding.

Reviewer 2: ```suggestion
The Kedro training materials assume a pre-requisite level of technical understanding.
```

> We assume understanding of these technologies and concepts during the training. In order to maximise your learning experience from the Kedro training, please review them before going through the training (there are external resource links for you to get a refresher on each topic).

Reviewer 2: ```suggestion
To optimise your experience and learn the most you can from the Kedro training, please review the following before your training session. We provide external resource links for each topic.
```

> - Intermediate Python knowledge

Reviewer 2: You say 'intermediate' here and then follow with 'basic python' which is a little confusing.

How about changing the link that says "Basic Python" to say "Python fundamentals" ?

  - [Basic Python](https://docs.python.org/3/tutorial/)
    - Functions, loops, conditional statements, IO operation
    - Common data structures: lists, dictionaries, tuples

>   - [Installing Python packages using `pip`](https://pip.pypa.io/en/stable/quickstart/) and [Dependency management with `requirements.txt`](https://pip.pypa.io/en/latest/user_guide/#requirements-files)

Reviewer 2: ```suggestion
  - [Installing Python packages using `pip`](https://pip.pypa.io/en/stable/quickstart/) 
  - [Dependency management with `requirements.txt`](https://pip.pypa.io/en/latest/user_guide/#requirements-files)
```

  - [Python modules](https://docs.python.org/3/tutorial/modules.html) (e.g how to use `__init__.py` and relative and absolute imports)
  - [Some familiarity with Python data science libraries](https://towardsdatascience.com/top-10-python-libraries-for-data-science-cd82294ec266), especially `Pandas`, `scikit-learn`, [Jupyter notebook/labs](https://www.dataquest.io/blog/jupyter-notebook-tutorial/), [iPython](https://www.codecademy.com/articles/how-to-use-ipython))

>   - [Using virtual environment](https://docs.python.org/3/tutorial/venv.html) (Anaconda, `venv`, `pipenv`)

Reviewer 2: ```suggestion
  - [Using a virtual environment](https://docs.python.org/3/tutorial/venv.html) (Anaconda, `venv`, `pipenv`)
```

- [Object-oriented programming](https://realpython.com/python3-object-oriented-programming/), especially Python classes, overriding class methods and inheritance.
- [Basic YAML syntax](https://yaml.org/)
- [Working with the command line](https://tutorial.djangogirls.org/en/intro_to_command_line/) (aka. cmd, CLI, prompt, console or terminal)
  -  `cd` to navigate directories
  -  `ls` to list files and directories
  -  [Executing a command and Python program from a command line](https://realpython.com/run-python-scripts/#how-to-run-python-scripts-using-the-command-line)

> Below is optional and it won't cause any issues of following the training materials if you don't know them (but Kedro integrates well with these technologies).

Reviewer 2: ```suggestion
The following lists technologies that Kedro integrates with, but it won't cause any issues if you are not yet familiar with them.
```

- [Version Control with Git](https://git-scm.com/doc)
- Cloud storage ([S3](https://aws.amazon.com/s3/), [Azure Blob](https://azure.microsoft.com/en-gb/services/storage/blobs/), [GCS](https://cloud.google.com/storage))
- [PySpark](https://spark.apache.org/docs/latest/api/python/index.html)
- [Docker](https://docs.docker.com/) for managing deployment
- [Airflow](https://airflow.apache.org/docs/stable/tutorial.html) for scheduling pipeline execution

# Checklist
Please use this checklist to make sure you have everything necessary to participate in the Kedro training.

> - [ ] You have Python 3 installed in your laptop.

Reviewer 2: ```suggestion
- [ ] You have Python 3 installed in your laptop.
```

Any particular version? Add a link or explanation of how to get it?

- [ ] You have Anaconda or an [alternative](https://github.com/quantumblacklabs/kedro-examples/blob/master/kedro-training/docs/02_virtual-environment.md) virtual environment manager

> - [ ] You have an editor installed for writing Python code

Reviewer 2: ...we recommend x, y or z ?? (with links)

Reviewer 2: I have just spotted that you have links in a section below -- so it's worth bringing that up here or linking to the section.

- [ ] You have a Terminal installed

> After you have installed above, please make sure that you are able to execute the following commands in your Terminal

Reviewer 2: ```suggestion
Having completed the above checklist, make sure that you are able to execute the following commands in your Terminal
```

# Training pre-requisites
This Kedro training materials assume the following pre-requisites for your technical understanding.

We assume understanding of these technologies and concepts during the training. In order to maximise your learning experience from the Kedro training, please review them before going through the training (there are external resource links for you to get a refresher on each topic).

- Intermediate Python knowledge

>   - [Basic Python](https://developers.google.com/edu/python)

Reviewer 3: Maybe link to python docs here too?

    - Functions, loops, conditional statements
    - Common data structures: lists, dictionaries, tuples

>     - [IO operation](https://realpython.com/working-with-files-in-python/) (reading and writing data)

Reviewer 3: do we actually use I/O operations anywhere in the tutorial (trying to remember), do we need to specify this? if we do, can we link to python I/O docs instead?

Author: I think they should know how to read/write data with Python even though we won't be touching directly, since all the datasets do this underneath. I moved IO operation in `Functions, loops, conditional statements, IO operration`, and removed the link :)

>   - [Importing Python packages using `pip`](https://www.w3schools.com/python/python_pip.asp) and [Dependency management with `requirements.txt`](https://note.nkmk.me/en/python-pip-install-requirements/)

Reviewer 3: Do you mean importing or installing python packages?

Can we link to the python/pip docs instead of w3 schools?

>   - [Working on multiple Python files in nested directory structures](https://drivendata.github.io/cookiecutter-data-science/)

Reviewer 3: I feel like this requirement is a little too specific 😅

Author: Agree, removed :)

# Training pre-requisites
This Kedro training materials assume the following pre-requisites for your technical understanding.

> We assume that you understand these technologies and we won't be teaching how to do any of them during the training. In order to maximise your learning experience from the Kedro training, please review them before the training if you are not familiar with any of them.

Reviewer 3: ```suggestion
We assume understanding of these technologies and concepts during the training. In order to maximise your learning experience from the Kedro training, please review them before going through the training.
```
maybe a little less confrontative 😅

- Intermediate Python knowledge
  - Basic Python
    - Functions
    - Classes
    - Loops
    - Conditional statements
    - Common data structures (lists, dictionaries, tuples)
    - IO operation (reading and writing data)
  - Importing Python packages using `pip`.
  - Dependency management with `requirements.txt`

>   - Working on multiple Python files in nested directory structures.

Reviewer 3: can we make this list shorter? I don't want to scare off beginners to Kedro :(

Author: Agree. I can make `Basic Python` part into one line, but it's not reducing anything... And I think most of them are really necessary to know to follow the training I guess? Happy for any suggestions for making this list look less scary :)

  - Python modules (e.g how to use `__init__.py` and relative and absolute imports)

>   - Familiarity with Python data science libraries (e.g `Pandas`, `PySpark`, `scikit-learn`, Jupyter notebook/labs, iPython)

Reviewer 3: Do we need to specify `pyspark` in here? I don't even know `pyspark` 👀

  - Using virtual environment (Anaconda, venv, pipenv)
- Object-oriented programming (overriding class methods, class inheritance).
- Basic YAML syntax
- Working with the command line (aka. cmd, CLI, prompt, console or terminal)
  -  `cd` to navigate directories
  -  `ls` to list files and directories
  -  Executing a command and Python program from a command line

Below is optional and it won't cause any issues of following the training materials if you don't know them (but Kedro integrates well with these technologies).
- Version Control with Git
- Cloud storage (S3, Azure Blob, GCP)
- PySpark
- Docker for managing deployment
- Airflow for scheduling pipeline execution

# Checklist
Please use this checklist to make sure you have everything necessary to participate in the Kedro training.

- [ ] You have Python 3 installed in your laptop.

> - [ ] You have Anaconda installed. An alternative to Anaconda can be found [here](https://github.com/quantumblacklabs/kedro-examples/blob/master/kedro-training/docs/02_virtual-environment.md)

Reviewer 3: ```suggestion
- [ ] You have Anaconda or an [alternative](https://github.com/quantumblacklabs/kedro-examples/blob/master/kedro-training/docs/02_virtual-environment.md) virtual environment manager
```

- [ ] You have an editor installed for writing Python code
- [ ] You have a Terminal installed

After you have installed above, please make sure that you are able to execute the following commands in your Terminal
- [ ]  `python --version` or `python3 --version` returns a correct Python version (either 3.6, 3.7 or 3.8).
- [ ] `git clone https://github.com/quantumblacklabs/kedro-examples` downloads this training Github repository (or [manually download a zip file from the link](https://stackoverflow.com/questions/2751227/how-to-download-source-in-zip-format-from-github)).

> - [ ] `conda create --name=training python=3.6 -y && conda activate training` creates an virtual environment called `training` and activate the environment.

Reviewer 3: ```suggestion
- [ ] `conda create --name=training python=3.6 -y && conda activate training` creates an virtual environment called `training` and activates the environment.
```
